### PR TITLE
feat: Replacing Juju bundle with Terraform module in Getting Started tutorial

### PR DESCRIPTION
**Pull request by Gmerold**
# Description

Updates the `Getting Started` tutorial to use Terraform modules instead of Juju bundle for Charmed 5G deployment.

Successful run depends on merging [this fix](https://github.com/canonical/terraform-juju-sdcore-k8s/pull/6).

# Checklist:

- [x] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that validate the behaviour of the software
- [x] I validated that new and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have bumped the version of the library

### CODE REVIEW

File: docs/tutorials/getting_started.md [hunk 1]
```diff
@@ -1,12 +1,15 @@
...
+the core and the Radio Access Network (RAN)) to simulate usage of this network. Both gNB Simulator 
+and SD-Core Router serve only demonstration purposes and shouldn't be part of production 
+deployments.
 
-To complete this tutorial, you will need a machine with the following
-requirements:
+To complete this tutorial, you will need a machine which meets the following requirements:
 
-- A recent `x86_64` CPU (Intel 4ᵗʰ generation or newer, or AMD Ryzen or newer)
+- A recent `x86_64` CPU (Intel 4ᵗʰ generation or newer or AMD Ryzen or newer)
```
Comment (markbeierl) [inline docs/tutorials/getting_started.md:12]: I would suggest the comma stays, as it is separating the grouping of "Intel or newer" from "AMD or newer"

Also, I just realized we did not specify the number of CPUs here, but that could be a different PR

Comment (Gmerold) [reply docs/tutorials/getting_started.md]: Done.
Also added info about required number of cores.

File: docs/tutorials/getting_started.md [hunk 2]
```diff
@@ -61,110 +65,197 @@ directory:
...
-```
-
-## 5. Configure the ingress
+## 6. Configure the ingress
 
 Configure Traefik to use an external hostname:
 
 ```console
-juju config traefik-k8s external_hostname=10.0.0.3.nip.io
+juju config traefik external_hostname=10.0.0.3.nip.io
```
Comment (markbeierl) [inline docs/tutorials/getting_started.md:243]: We seem to have a race condition problem here.  In my case, traefik took .4, not .3:
```
sdcore                          amf-external                         LoadBalancer   10.152.183.68    10.0.11.2     38412:32503/SCTP              35m
sdcore                          upf-external                         LoadBalancer   10.152.183.34    10.0.11.3     8805:31595/UDP                33m
sdcore                          traefik                              LoadBalancer   10.152.183.70    10.0.11.4     80:32753/TCP,443:31561/TCP    36m
```

Note that I used 10.0.11.2-10.0.11.4 as the range due to 10.0.0.0 being my home subnet

File: docs/tutorials/getting_started.md [hunk 3]
```diff
@@ -61,110 +65,197 @@ directory:
...
 ```console
-juju run traefik-k8s/0 show-proxied-endpoints
+juju run traefik/0 show-proxied-endpoints
 ```
 
 The output should be `http://core-nms.10.0.0.3.nip.io/`. Navigate to this address in your browser.
 
 
-## 6. Configure the 5G core network through the Network Management System
+## 7. Configure the 5G core network through the Network Management System
```
[the remark this answers is not in the document]
Comment (Gmerold) [reply docs/tutorials/getting_started.md:258]: Nice catch :) 
Done.

**Review by markbeierl — COMMENTED** [review]
I accidentally committed and pushed directly here. I meant to make suggestions of changes, not commit, but either way, I just found and resolved 2 more `core` vs. `sdcore` changes

**Comment by markbeierl** [comment]
Do we want to update the picture as well? 
image

**Comment by markbeierl** [comment]
Re-running Terraform will revert the change for the `external_hostname`:
```
Terraform will perform the following actions:

  # module.sdcore.module.traefik.juju_application.traefik-k8s will be updated in-place
  ~ resource "juju_application" "traefik-k8s" {
      ~ config      = {
          - "external_hostname" = "10.0.11.4.nip.io" -> null
            # (1 unchanged element hidden)
        }
        id          = "sdcore:traefik"
        name        = "traefik"
      + principal   = (known after apply)
        # (4 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }
```

We should change the tutorial to make this modification via TF instead of directly against Juju

**Comment by Gmerold** [comment]
> Do we want to update the picture as well? image

Done.

**Comment by Gmerold** [comment]
> Re-running Terraform will revert the change for the `external_hostname`:
> 
> ```
> Terraform will perform the following actions:
> 
>   # module.sdcore.module.traefik.juju_application.traefik-k8s will be updated in-place
>   ~ resource "juju_application" "traefik-k8s" {
>       ~ config      = {
>           - "external_hostname" = "10.0.11.4.nip.io" -> null
>             # (1 unchanged element hidden)
>         }
>         id          = "sdcore:traefik"
>         name        = "traefik"
>       + principal   = (known after apply)
>         # (4 unchanged attributes hidden)
> 
>         # (1 unchanged block hidden)
>     }
> ```
> 
> We should change the tutorial to make this modification via TF instead of directly against Juju

True. Done.

File: docs/tutorials/getting_started.md [hunk 4]
```diff
@@ -1,12 +1,16 @@
 # Getting started
 
-In this tutorial, we will deploy and run the SD-Core 5G core network using Juju. We will also
-deploy a radio and cellphone simulator to simulate usage of this network.
+In this tutorial, we will deploy and run the SD-Core 5G core network using Juju and Terraform.
+As part of this tutorial, we will also deploy additional components (gNB Simulator - a 5G radio 
+and a cellphone simulator, SD-Core Router - a software router facilitating communication between
+the core and the Radio Access Network (RAN)) to simulate usage of this network. Both gNB Simulator 
+and SD-Core Router serve only demonstration purposes and shouldn't be part of production
```
Comment (patriciareinoso) [inline docs/tutorials/getting_started.md:7]:
```suggestion
and SD-Core Router serve only for demonstration purposes and shouldn't be part of production 
```

File: docs/tutorials/getting_started.md [hunk 5]
```diff
@@ -61,118 +65,229 @@ directory:
...
+From your terminal, install Terraform.
 
-Create a Juju model named `core`:
+```console
+sudo snap install terraform --classic
+```
+
+## 4. Create Terraform module
+
+On the host machine create a new folder called `terraform`:
```
Comment (patriciareinoso) [inline docs/tutorials/getting_started.md:78]: I think I would say `directory` instead of `folder` but I don't know what the right term to use in English is.
(Same in line 84, 169)

File: docs/tutorials/getting_started.md (hunk 4 above)
Comment (Gmerold) [reply docs/tutorials/getting_started.md:7]: I think my version was ok: https://dictionary.cambridge.org/dictionary/english/serve-a-purpose

File: docs/tutorials/getting_started.md (hunk 5 above)
Comment (Gmerold) [reply docs/tutorials/getting_started.md]: I think they essentially mean the same. Purists might say that the folder is a graphical representation of a directory, but to me it really boils down to the Potahto-Potato problem :)
Changed to `directory` ;)

File: docs/tutorials/getting_started.md (hunk 4 above)
Comment (markbeierl) [reply docs/tutorials/getting_started.md:7]: As the SD-Core Router is a singular object, the verb needs to be `serves`

File: docs/tutorials/getting_started.md [hunk 6]
```diff
@@ -61,118 +65,229 @@ directory:
...
-Configure Traefik to use an external hostname:
+Apply new configuration:
 
 ```console
-juju config traefik-k8s external_hostname=10.0.0.3.nip.io
+terraform apply -auto-approve
 ```
 
-Here, replace `10.0.0.3` with the Application IP address of the `traefik-k8s` application. You can find it by running `juju status traefik-k8s`.
+```{note}
```
Comment (markbeierl) [inline docs/tutorials/getting_started.md:267]: I think this should be an unconditional step before the step to edit the `main.tf` file.  Like:

* run `juju status traefik`
* note the IP address
* edit the `main.tf`

Comment (Gmerold) [reply docs/tutorials/getting_started.md]: Yeah, makes sense.
Done.

File: docs/tutorials/getting_started.md [hunk 7]
```diff
@@ -61,118 +65,232 @@ directory:
...
+EOF
 ```
 
-Deploy the `sdcore-k8s` charm bundle:
+Create a Terraform module containing the SD-Core 5G core network, 5G radio and a cellphone
+simulator and a router:
 
 ```console
-juju deploy sdcore-k8s --trust --channel=beta
+cd terraform
```
Comment (markbeierl) [inline docs/tutorials/getting_started.md:104]: duplicate of L#87, and gives "no such file"

File: docs/tutorials/getting_started.md [hunk 8]
```diff
@@ -61,118 +65,232 @@ directory:
...
+Initialize Juju Terraform provider:
+
+```console
+terraform init
+```
+
+Deploy 5G network.
+
+```console
+terraform apply -auto-approve
```
Comment (markbeierl) [inline docs/tutorials/getting_started.md:183]: I also get `This object does not have an attribute named "database_endpoint".`

Perhaps this is due to the change in auth databases, @gatici ?

Comment (Gmerold) [reply docs/tutorials/getting_started.md:182]: Yes, that's exactly the reason. I've updated the modules, so it should be good again.

**Review by markbeierl — APPROVED** [review]

**Review by patriciareinoso — APPROVED** [review]
